internal/initdata/initdata.go

```diff
@@ -125,10 +128,60 @@ func (i *Initdata) EncodeKataAnnotation() (string, string, error) {
 		return "", "", err
 	}
 
-	return encoded, digest, nil
+	// TODO(burgerdev): why is the digest a hex string?
+	return encoded, hex.EncodeToString(digest), nil
 }
 
-func (i *Initdata) digest(digestible []byte) (string, error) {
+// FromDevice reads initdata and its digest from a block device prepared by the Kata runtime.
+func FromDevice(devicePath string) (*Initdata, []byte, error) {
+	f, err := os.Open(devicePath)
+	if err != nil {
+		return nil, nil, err
+	}
+	buf := make([]byte, 8)
+	_, err = f.ReadAt(buf, 0)
+	if err != nil {
+		return nil, nil, fmt.Errorf("reading magic number: %w", err)
+	}
+	const magic = "initdata"
+	if slices.Compare(buf, []byte(magic)) != 0 {
+		return nil, nil, fmt.Errorf("%w: expected %x, got %x", errWrongMagic, magic, buf)
+	}
+	buf = make([]byte, 8)
+	_, err = f.ReadAt(buf, 8)
+	if err != nil {
+		return nil, nil, fmt.Errorf("reading magic number: %w", err)
+	}
+	size := binary.LittleEndian.Uint64(buf)
+	const maxSize = 128 * 1024 * 1024 * 1024
+	if size > maxSize {
+		return nil, nil, fmt.Errorf("%w: expected at most 128MiB, got %d byte", errTooLarge, size)
+	}
+	buf = make([]byte, size)
+
+	_, err = f.ReadAt(buf, 16)
+	if err != nil {
+		return nil, nil, fmt.Errorf("reading initdata blob: %w", err)
+	}
+
+	decompressed, err := decompress(buf)
+	if err != nil {
+		return nil, nil, fmt.Errorf("unzipping initdata: %w", err)
+	}
+
+	var i Initdata
+	if err := toml.Unmarshal(decompressed, &i); err != nil {
+		return nil, nil, fmt.Errorf("parsing TOML: %w", err)
+	}
+	digest, err := i.digest(decompressed)
+	if err != nil {
+		return nil, nil, fmt.Errorf("calculating digest: %w", err)
+	}
+
+	return &i, digest, nil
```

Comment:
This is almost 1:1 the same as `DecodeKataAnnotation`, minus verifying that a known algorithm is used. A shared helper would seem to make sense.